{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |default_exp sample_manager\n",
    "# You need this at the top of every notebook you want turned into a module, the name your provide will determine the module name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "# That export there, it makes sure this code goes into the module.\n",
    "\n",
    "# standard libs\n",
    "import os\n",
    "import re\n",
    "\n",
    "# Common to template\n",
    "# add into settings.ini, requirements, package name is python-dotenv, for conda build ensure `conda config --add channels conda-forge`\n",
    "import dotenv  # for loading config from .env files, https://pypi.org/project/python-dotenv/\n",
    "import envyaml  # Allows to loads env vars into a yaml file, https://github.com/thesimj/envyaml\n",
    "import fastcore  # To add functionality related to nbdev development, https://github.com/fastai/fastcore/\n",
    "from fastcore import (\n",
    "    test,\n",
    ")\n",
    "from fastcore.script import (\n",
    "    call_parse,\n",
    ")  # for @call_parse, https://fastcore.fast.ai/script\n",
    "import json  # for nicely printing json and yaml\n",
    "from fastcore import test\n",
    "from fastcore.script import call_parse\n",
    "from ssi_analysis_utility import (\n",
    "    core,\n",
    ")\n",
    "from pathlib import Path #to be able write :Path in cli function\n",
    "\n",
    "# Project specific libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To change working directory when running notebook only. DO NOT EXPORT\n",
    "\n",
    "os.chdir(core.PROJECT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample manager\n",
    "\n",
    "\n",
    "## sample_data\n",
    "\n",
    "The sample data class is simple. It initiates an object with the following attributes:\n",
    "\n",
    "- sample_name: an identifier for the sample\n",
    "- assembly_file: Path to a genome assembly in fasta format, i.e. \"/path/to/genome_assembly.fasta\"\n",
    "- Illumina_read_files: Paths to R1 and R2 from Illumina paired end sequencing in a comma-separated string, i.e. \"/path/to/R1.fastq.gz,/path/to/R2.fastq.gz\"\n",
    "- Nanopore_read_file: Path to a file containing long read sequencing data in fastq format, i.e. \"/path/to/Nanopore_read_data.fastq.gz\"\n",
    "- metadata: A dictionary containing metadata read from a provided tsv-file. Will also contain the above values (sample_name, paths to data files)\n",
    "- input_folder: If provided. Needed when loading multiple samples at once using the input_manager class (see below)\n",
    "\n",
    "\n",
    "## input_manager\n",
    "\n",
    "The input_manager class is a way to initiate sample_data class instances in three different ways:\n",
    "\n",
    "1. Initiating just a single sample_data class instance by providing a sample name and paths to read/assembly files\n",
    "2. Initiating class instances for all samples in a specific input folder. Assembly files, Illumina read files and Nanopore read files are auto detected and matched by sample name.\n",
    "3. Initiating class instances for samples in a provided metadatasheet sheet with paths to those samples.\n",
    "\n",
    "For all three options metadata can be provided in tsv-format.\n",
    "\n",
    "## Usage\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "\n",
    "class sample_data:\n",
    "\n",
    "    species = \"Unspecified\"\n",
    "\n",
    "    def __init__(self,attributes,input_folder):\n",
    "        attributes = attributes.copy()\n",
    "        try:\n",
    "            self.input_folder = os.path.abspath(input_folder)\n",
    "        except TypeError:\n",
    "            self.input_folder = None\n",
    "        try:\n",
    "            self.sample_name: str = attributes['sample_name']\n",
    "        except KeyError:\n",
    "            self.sample_name = \"sample\"\n",
    "        try:\n",
    "            self.assembly_file = get_abs_file_path(self.input_folder,attributes['assembly_file'])\n",
    "        except KeyError:\n",
    "            self.assembly_file = None\n",
    "        if \"Illumina_read_files\" in attributes and attributes[\"Illumina_read_files\"] is not None:\n",
    "            if isinstance(attributes['Illumina_read_files'], str):\n",
    "                Illumina_read_files = attributes['Illumina_read_files'].split(\",\")\n",
    "            elif isinstance(attributes['Illumina_read_files'], list):\n",
    "                Illumina_read_files = attributes['Illumina_read_files']\n",
    "            self.Illumina_read_files = [get_abs_file_path(self.input_folder,Illumina_read_files[0]),get_abs_file_path(self.input_folder,Illumina_read_files[1])]\n",
    "        else:\n",
    "            self.Illumina_read_files = None\n",
    "        try:\n",
    "            self.Nanopore_read_file: Path = get_abs_file_path(self.input_folder,attributes['Nanopore_read_file'])\n",
    "        except KeyError:\n",
    "            self.Nanopore_read_file = None\n",
    "        self.metadata: dict = attributes\n",
    "\n",
    "\n",
    "\n",
    "def get_abs_file_path(input_folder,file_path):\n",
    "    if file_path is not None:\n",
    "        if input_folder:\n",
    "            file_path = os.path.abspath(os.path.join(input_folder,file_path))\n",
    "        else:\n",
    "            file_path = os.path.abspath(file_path)\n",
    "        if not os.path.exists(file_path):\n",
    "            print(f\"WARNING: {file_path} does not exist\")\n",
    "    return(file_path)\n",
    "\n",
    "#def load_samplesheet_tsv(samplesheet):\n",
    "#    df = core.get_samplesheet({\"path\": samplesheet})\n",
    "#    df.set_index('sample_name', inplace=True)\n",
    "#    df_dict = df.to_dict(\"index\")\n",
    "#    return(df_dict)\n",
    "\n",
    "\n",
    "def get_single_sample_attributes(input_config):\n",
    "    file_paths = dict((k, input_config[k]) for k in [\"sample_name\",\"assembly_file\",\"Illumina_read_files\",\"Nanopore_read_file\"] if k in input_config)\n",
    "    if not \"sample_name\" in input_config or input_config[\"sample_name\"] is None:\n",
    "        if \"assembly_file\" in input_config and input_config[\"assembly_file\"] is not None:\n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"assembly_fasta_file\"],os.path.basename(input_config[\"assembly_file\"])).group(\"sample_name\")\n",
    "        elif \"Illumina_read_files\" in input_config and input_config[\"Illumina_read_files\"] is not None:\n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"Illumina_read_file\"],os.path.basename(input_config[\"Illumina_read_files\"].split(',')[0])).group(\"sample_name\")\n",
    "        elif \"Nanopore_read_file\" in input_config and input_config[\"Nanopore_read_file\"] is not None:\n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"assembly_fasta_file\"],os.path.basename(input_config[\"Nanopore_read_file\"])).group(\"sample_name\")\n",
    "    if input_config[\"samplesheet\"] is not None:\n",
    "        df = core.get_samplesheet({\"path\": input_config[\"samplesheet\"]})\n",
    "        try:\n",
    "            df.set_index('sample_name', inplace=True)\n",
    "            df_dict = df.to_dict(\"index\")\n",
    "            if input_config[\"sample_name\"] is not None and input_config[\"sample_name\"] in df_dict:\n",
    "                attributes = df_dict[input_config[\"sample_name\"]]\n",
    "            else:\n",
    "                attributes = df_dict[list(df_dict)[0]]\n",
    "                print(\"Sample name not provided or not found in metadata. Using first row as metadata input\")\n",
    "        except KeyError:\n",
    "            df_list = df.to_dict(\"records\")\n",
    "            attributes = df_list[0]\n",
    "            print(\"No sample_name column in metadata. Using first row as metadata input\")\n",
    "    else:\n",
    "        attributes = {}\n",
    "    attributes.update(file_paths)\n",
    "    return(attributes)\n",
    "\n",
    "\n",
    "\n",
    "class input_manager:\n",
    "    def __init__(self,input_config):\n",
    "        try:\n",
    "            self.base_input_folder: Path = os.path.abspath(input_config[\"input_folder\"])\n",
    "        except:\n",
    "            self.base_input_folder: Path = os.path.abspath(\"./\")\n",
    "        self.get_samplesheet(input_config)\n",
    "        if input_config[\"load_from_samplesheet\"]:\n",
    "            if not input_config[\"input_folder\"] or input_config[\"input_folder\"] is None:\n",
    "                self.base_input_folder = os.path.dirname(self.metadata_file)\n",
    "            self.add_samples_from_samplesheet()\n",
    "        else:\n",
    "            self.add_samples_from_folder(input_config)\n",
    "    \n",
    "    ### TODO: figure out how to deal with duplicate sample names. Just a warning?\n",
    "\n",
    "    def get_samplesheet(self,input_config):\n",
    "        if \"samplesheet\" in input_config and input_config[\"samplesheet\"] is not None:\n",
    "            self.metadata_file = os.path.abspath(input_config[\"samplesheet\"])\n",
    "            self.metadata = self.get_metadata_from_samplesheet()\n",
    "        else:\n",
    "            self.metadata_file = None\n",
    "            self.metadata = None\n",
    "\n",
    "\n",
    "    def init_sample(self,attributes):\n",
    "        sample = sample_data(attributes,self.base_input_folder)\n",
    "        return(sample)\n",
    "              \n",
    "    def add_samples(self,file_paths):\n",
    "        self.samples = []\n",
    "        for sample_name in file_paths:\n",
    "            attributes = file_paths[sample_name].copy()\n",
    "            attributes[\"sample_name\"] = sample_name\n",
    "            self.samples.append(self.init_sample(attributes))\n",
    "        return(file_paths)\n",
    "\n",
    "    def add_samples_from_samplesheet(self):\n",
    "        file_paths = self.get_metadata_from_samplesheet()\n",
    "        self.add_samples(file_paths)\n",
    "    \n",
    "    def add_samples_from_folder(self,input_config):\n",
    "        file_paths = self.get_input_from_folder(input_config)\n",
    "        self.add_samples(file_paths)\n",
    "\n",
    "    def get_input_from_folder(self,input_config):\n",
    "        illumina_regex = re.compile(input_config[\"file_patterns\"][\"Illumina_read_file\"])\n",
    "        nanopore_regex = re.compile(input_config[\"file_patterns\"][\"Nanopore_read_file\"])\n",
    "        assembly_regex = re.compile(input_config[\"file_patterns\"][\"assembly_fasta_file\"])\n",
    "        input_folder = self.base_input_folder\n",
    "        file_paths = {}\n",
    "        files = os.listdir(input_folder)\n",
    "        for file in files:\n",
    "            sample_name = False\n",
    "            if file.endswith((\".fa\",\".fasta\",\".fna\")):\n",
    "                sample_name = re.match(assembly_regex,file).group(\"sample_name\")\n",
    "                if sample_name in file_paths:\n",
    "                    file_paths[sample_name][\"assembly_file\"] = file\n",
    "                else:\n",
    "                    file_paths[sample_name] = {\"assembly_file\": file}\n",
    "            elif file.endswith(\".fastq.gz\"):\n",
    "                re_match = re.match(illumina_regex,file)\n",
    "                if re_match:\n",
    "                    sample_name = re_match.group(\"sample_name\")\n",
    "                    read_number = re_match.group(\"paired_read_number\")\n",
    "                    if read_number == \"R1\":\n",
    "                        if sample_name in file_paths:\n",
    "                            if \"Illumina_read_files\" in file_paths[sample_name]:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file + \",\" + file_paths[sample_name][\"Illumina_read_files\"]\n",
    "                            else:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Illumina_read_files\": file}\n",
    "                    elif read_number == \"R2\":\n",
    "                        if sample_name in file_paths:\n",
    "                            if \"Illumina_read_files\" in file_paths[sample_name]:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file_paths[sample_name][\"Illumina_read_files\"] + \",\" + file\n",
    "                            else:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Illumina_read_files\": file}\n",
    "                \n",
    "                #### TODO: Logic for nanopore read data, need file naming conventions\n",
    "                else:\n",
    "                    re_match = re.match(nanopore_regex,file)\n",
    "                    if re_match:\n",
    "                        sample_name = re_match.group(\"sample_name\")\n",
    "                        if sample_name in file_paths:\n",
    "                            file_paths[sample_name][\"Nanopore_read_file\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Nanopore_read_file\": file}\n",
    "        ### update attributes with metadata from samplesheet if provided\n",
    "        if self.metadata_file is not None:\n",
    "            for sample_name in file_paths:\n",
    "                if sample_name in self.metadata:\n",
    "                    file_paths_updated = self.metadata[sample_name].copy()\n",
    "                    file_paths_updated.update(file_paths[sample_name])\n",
    "                    file_paths[sample_name] = file_paths_updated\n",
    "        return(file_paths)\n",
    "    \n",
    "    def get_metadata_from_samplesheet(self):\n",
    "        df = core.get_samplesheet({\"path\": self.metadata_file})\n",
    "        df.set_index('sample_name', inplace=True)\n",
    "        df_dict = df.to_dict(\"index\")\n",
    "        return(df_dict)\n",
    "\n",
    "    def __iter__(self):\n",
    "        for sample in self.samples:\n",
    "            yield(sample)\n",
    "\n",
    "    \n",
    "    def __len__(self):\n",
    "        return(len(self.samples))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "def single_test():\n",
    "    example_sample = sample_data({#\"sample_name\":\"GAS-2022-1029\",\n",
    "                                       \"assembly_file\":\"examples/GAS-2022-1029.fasta\",\n",
    "                                       \"Illumina_read_files\":[\"examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz\",\"examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz\"]},\n",
    "                                       input_folder = False,\n",
    "                                       )\n",
    "    assert(len(example_sample.Illumina_read_files) == 2)\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "def single_test_assembly():\n",
    "    example_sample = sample_data({\"assembly_file\":\"examples/GAS-2022-1029.fasta\"},\n",
    "                                       input_folder = False,\n",
    "                                       )\n",
    "    assert(not(example_sample.Illumina_read_files))\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "def single_test_input_folder():\n",
    "    config = core.get_config()\n",
    "    example_sample = sample_data({\"sample_name\":\"GAS-2024-0773\",\n",
    "                                       \"assembly_file\":\"GAS-2024-0773.fasta\",\n",
    "                                       \"Illumina_read_files\":[\"GAS-2024-0773_S35_L555_R1_001.fastq.gz\",\"GAS-2024-0773_S35_L555_R2_001.fastq.gz\"]},\n",
    "                                       input_folder = \"examples\",\n",
    "                                       )\n",
    "    assert(example_sample.sample_name == \"GAS-2024-0773\")\n",
    "    assert(len(example_sample.Illumina_read_files) == 2)\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "\n",
    "def test_folder():\n",
    "    config = core.get_config()\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_folder\"] = True\n",
    "    input_config[\"input_folder\"] = \"examples/\"\n",
    "    #input_config[\"samplesheet\"] = \"samplesheet.tsv\"\n",
    "    test = input_manager(input_config)\n",
    "    for x in test:\n",
    "        print(x.__dict__)\n",
    "\n",
    "def test_folder_with_metadata():\n",
    "    config = core.get_config()\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_folder\"] = True\n",
    "    input_config[\"input_folder\"] = \"examples/\"\n",
    "    input_config[\"samplesheet\"] = \"examples/samplesheet.tsv\"\n",
    "    test = input_manager(input_config)\n",
    "    for x in test:\n",
    "        print(x.__dict__)\n",
    "\n",
    "def samplesheet_test():\n",
    "    config = core.get_config()\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_samplesheet\"] = True\n",
    "    input_config[\"samplesheet\"] = \"examples/samplesheet.tsv\"\n",
    "    test = input_manager(input_config)\n",
    "    for x in test:\n",
    "        print(x.metadata[\"Illumina_read_files\"])\n",
    "        print(x.metadata[\"assembly_file\"])\n",
    "        print(x.metadata[\"sample_name\"])\n",
    "        print(x.assembly_file)\n",
    "        print(x.Illumina_read_files)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': None, 'sample_name': 'sample', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', 'examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz']}}\n",
      "{'input_folder': None, 'sample_name': 'sample', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'examples/GAS-2022-1029.fasta'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'sample_name': 'GAS-2024-0773', 'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': ['GAS-2024-0773_S35_L555_R1_001.fastq.gz', 'GAS-2024-0773_S35_L555_R2_001.fastq.gz']}}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "single_test()\n",
    "single_test_assembly()\n",
    "single_test_input_folder()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2023-0253', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_nanopore.fastq.gz', 'metadata': {'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'sample_name': 'GAS-2023-0253'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2022-1029', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_nanopore.fastq.gz', 'metadata': {'assembly_file': 'GAS-2022-1029.fasta', 'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'sample_name': 'GAS-2022-1029'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': 'GAS-2024-0773_S35_L555_R1_001.fastq.gz,GAS-2024-0773_S35_L555_R2_001.fastq.gz', 'sample_name': 'GAS-2024-0773'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2024-0665', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0665.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0665.fasta', 'sample_name': 'GAS-2024-0665'}}\n"
     ]
    }
   ],
   "source": [
    "test_folder()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2023-0253', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_nanopore.fastq.gz', 'metadata': {'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1UK', 'notes': nan, 'sample_name': 'GAS-2023-0253'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2022-1029', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_nanopore.fastq.gz', 'metadata': {'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'assembly_file': 'GAS-2022-1029.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'speC positive', 'sample_name': 'GAS-2022-1029'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': 'GAS-2024-0773_S35_L555_R1_001.fastq.gz,GAS-2024-0773_S35_L555_R2_001.fastq.gz', 'sample_name': 'GAS-2024-0773'}}\n",
      "{'input_folder': '/Users/b246838/Documents/git.repositories/streptofile/examples', 'sample_name': 'GAS-2024-0665', 'assembly_file': '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2024-0665.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0665.fasta', 'sample_name': 'GAS-2024-0665'}}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "test_folder_with_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz\n",
      "GAS-2022-1029.fasta\n",
      "GAS-2022-1029\n",
      "/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029.fasta\n",
      "['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz']\n",
      "GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz\n",
      "GAS-2023-0253.fasta\n",
      "GAS-2023-0253\n",
      "/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253.fasta\n",
      "['/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R1_001.fastq.gz', '/Users/b246838/Documents/git.repositories/streptofile/examples/GAS-2023-0253_S74_L555_R2_001.fastq.gz']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "samplesheet_test()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "# This is included at the end to ensure when you run through your notebook the code is also transferred to the module and isn't just a notebook\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sample_name': 'GAS-2022-1029',\n",
       " 'assembly_file': None,\n",
       " 'Illumina_read_files': 'examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz,examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz',\n",
       " 'Nanopore_read_file': None}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "config = core.get_config()\n",
    "input_config = config[\"input_manager\"]\n",
    "input_config.update({#\"assembly_file\":\"examples/GAS-2022-1029.fasta\",\n",
    "                     \"Illumina_read_files\": \"examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz,examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz\"})\n",
    "get_single_sample_attributes(input_config)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
