{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |default_exp sample_manager\n",
    "# You need this at the top of every notebook you want turned into a module, the name your provide will determine the module name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "# That export there, it makes sure this code goes into the module.\n",
    "\n",
    "# standard libs\n",
    "import os\n",
    "import re\n",
    "\n",
    "# Common to template\n",
    "# add into settings.ini, requirements, package name is python-dotenv, for conda build ensure `conda config --add channels conda-forge`\n",
    "import dotenv  # for loading config from .env files, https://pypi.org/project/python-dotenv/\n",
    "import envyaml  # Allows to loads env vars into a yaml file, https://github.com/thesimj/envyaml\n",
    "import fastcore  # To add functionality related to nbdev development, https://github.com/fastai/fastcore/\n",
    "from fastcore import (\n",
    "    test,\n",
    ")\n",
    "from fastcore.script import (\n",
    "    call_parse,\n",
    ")  # for @call_parse, https://fastcore.fast.ai/script\n",
    "import json  # for nicely printing json and yaml\n",
    "from fastcore import test\n",
    "from fastcore.script import call_parse\n",
    "from ssi_analysis_utility import (\n",
    "    core,\n",
    ")\n",
    "from pathlib import Path #to be able write :Path in cli function\n",
    "\n",
    "# Project specific libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To change working directory when running notebook only. DO NOT EXPORT\n",
    "\n",
    "os.chdir(core.PROJECT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample Manager\n",
    "\n",
    "The Sample Manager module provides tools for managing and initializing sample data from Illumina and Nanopore sequencing. It includes two main classes: `sample_data` for individual sample attributes and `input_manager` for creating and handling multiple samples efficiently.\n",
    "\n",
    "---\n",
    "\n",
    "## `sample_data` Class\n",
    "\n",
    "The `sample_data` class is used to represent and store essential information for the samples. This class creates an object with the following attributes:\n",
    "\n",
    "- **sample_name**: Identifier for the sample, e.g unique name or code.\n",
    "- **assembly_file**: Path to the genome assembly file in FASTA format, e.g., `\"/path/to/genome_assembly.fasta\"`.\n",
    "- **Illumina_read_files**: Paths to paired-end sequencing files (R1 and R2) from Illumina sequencing, formatted as a comma-separated string, e.g., `\"/path/to/R1.fastq.gz,/path/to/R2.fastq.gz\"`.\n",
    "- **Nanopore_read_file**: Path to a file containing long-read sequencing data in FASTQ format, e.g., `\"/path/to/Nanopore_read_data.fastq.gz\"`.\n",
    "- **metadata**: A dictionary containing metadata loaded from a TSV file, including `sample_name` and paths to data files.\n",
    "- **input_folder**: Optional path used when loading multiple samples at once with the `input_manager` class (see below).\n",
    "\n",
    "---\n",
    "\n",
    "## `input_manager` Class\n",
    "\n",
    "The `input_manager` class is designed to initiate `sample_data` class instances. It provides three different ways to initialize samples:\n",
    "\n",
    "1. **Single Sample Initialization**:\n",
    "    - Creates a single `sample_data` instance by directly providing the sample name and paths to the read/assembly files.\n",
    "\n",
    "2. **Folder-Based Initialization**:\n",
    "    - Initializes `sample_data` instances for all samples in a specified input folder.\n",
    "    - The class automatically identifies and matches files by sample name for assembly files, Illumina read files, and Nanopore read files.\n",
    "\n",
    "3. **Metadata Sheet-Based Initialization**:\n",
    "    - Initializes `sample_data` instances for all samples listed in a provided metadata sheet.\n",
    "    - This metadata sheet should contain paths to each sampleâ€™s relevant files.\n",
    "\n",
    "For all three options, additional metadata can be provided in TSV format and will be included in the `metadata` attribute of each `sample_data` instance.\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "\n",
    "class sample_data:\n",
    "    \"\"\"\n",
    "    Class to handle sample data\n",
    "    \"\"\"\n",
    "\n",
    "    species = \"Unspecified\"\n",
    "\n",
    "    def __init__(self,\n",
    "                 attributes,\n",
    "                 input_folder):\n",
    "        \"\"\"\n",
    "        Initialize the class \n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        attributes : dict \n",
    "            dictionary containing sample information \n",
    "        input_folder : string\n",
    "            Path to input folder e.g reads\n",
    "        \"\"\"\n",
    "        attributes = attributes.copy()\n",
    "        #  Try to get the absolute path of the input folder\n",
    "        try:\n",
    "            self.input_folder = os.path.abspath(input_folder)\n",
    "        except TypeError:\n",
    "            # Set to None if there's an issue with input_folder\n",
    "            self.input_folder = None \n",
    "        \n",
    "        # Try to retrieve the sample name from attributes\n",
    "        try:\n",
    "            self.sample_name: str = attributes['sample_name']\n",
    "        except KeyError:\n",
    "            self.sample_name = \"sample\" \n",
    "\n",
    "        try:\n",
    "            # Try to retrieve the assembly file path and make it absolute\n",
    "            self.assembly_file = get_abs_file_path(self.input_folder,\n",
    "                                                   attributes['assembly_file'])\n",
    "        except KeyError:\n",
    "            # If 'assembly_file' key doesn't exist\n",
    "            self.assembly_file = None\n",
    "\n",
    "        # Try to retrieve Illumina read file\n",
    "        if \"Illumina_read_files\" in attributes and attributes[\"Illumina_read_files\"] is not None:\n",
    "            if isinstance(attributes['Illumina_read_files'], str):\n",
    "                Illumina_read_files = attributes['Illumina_read_files'].split(\",\")\n",
    "            elif isinstance(attributes['Illumina_read_files'], list):\n",
    "                Illumina_read_files = attributes['Illumina_read_files']\n",
    "            self.Illumina_read_files = [get_abs_file_path(self.input_folder,\n",
    "                                                          Illumina_read_files[0]), # First read\n",
    "                                        get_abs_file_path(self.input_folder,\n",
    "                                                          Illumina_read_files[1])] # Second read\n",
    "        else:\n",
    "            self.Illumina_read_files = None # If 'Illumina_read_files' key doesn't exist\n",
    "\n",
    "        try:\n",
    "            # Try to retrieve Nanopore read file and make it absolute\n",
    "            self.Nanopore_read_file: Path = get_abs_file_path(self.input_folder,\n",
    "                                                              attributes['Nanopore_read_file'])\n",
    "        except KeyError:\n",
    "            self.Nanopore_read_file = None # If 'Nanopore_read_file' key doesn't exist\n",
    "        # Store all attributes as metadata\n",
    "        self.metadata: dict = attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "class input_manager:\n",
    "    \"\"\"\n",
    "    class to manage input configurations and sample initialization\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    def __init__(self,\n",
    "                 input_config):\n",
    "        \"\"\"\n",
    "        Initialize the class \n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        input_config : yaml file\n",
    "            Configuration file with options to handle samples\n",
    "        \"\"\" \n",
    "        \n",
    "\n",
    "        try:\n",
    "            # Try to set the base output folder; default to './' if not provided\n",
    "            self.base_input_folder: Path = os.path.abspath(input_config[\"input_folder\"])\n",
    "        except:\n",
    "            self.base_input_folder: Path = os.path.abspath(\"./\")\n",
    "        self.get_samplesheet(input_config)\n",
    "\n",
    "        # Decide on loading method based on configuration settings\n",
    "        if input_config[\"load_from_samplesheet\"]:\n",
    "            # If loading from samplesheet, update base input folder if not specified\n",
    "            if not input_config[\"input_folder\"] or input_config[\"input_folder\"] is None:\n",
    "                self.base_input_folder = os.path.dirname(self.metadata_file)\n",
    "            self.add_samples_from_samplesheet()\n",
    "        else:\n",
    "            # Load samples directly from specified folder\n",
    "            self.add_samples_from_folder(input_config)\n",
    "    \n",
    "    ### TODO: figure out how to deal with duplicate sample names. Just a warning?\n",
    "\n",
    "    def get_samplesheet(self,\n",
    "                        input_config):\n",
    "        \"\"\"\n",
    "        Method to retrieve the samplesheet from the configuration file\n",
    "        and use to update metadata\n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        input_config : yaml file\n",
    "            Configuration file with options to handle samples\n",
    "        \"\"\"\n",
    "        # Check if there is a samplesheet provided in the configuration\n",
    "        if \"samplesheet\" in input_config and input_config[\"samplesheet\"] is not None:\n",
    "            self.metadata_file = os.path.abspath(input_config[\"samplesheet\"])\n",
    "            self.metadata = self.get_metadata_from_samplesheet() # Get metadata\n",
    "        else:\n",
    "            # No information\n",
    "            self.metadata_file = None\n",
    "            self.metadata = None\n",
    "\n",
    "\n",
    "    def init_sample(self,\n",
    "                    attributes):\n",
    "        \"\"\"\n",
    "        Method to initialize a single sample object from sample_data class\n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        attributes : dict \n",
    "            dictionary containing sample information \n",
    "        \"\"\"\n",
    "        \n",
    "        sample = sample_data(attributes,\n",
    "                             self.base_input_folder)\n",
    "        return(sample)\n",
    "              \n",
    "    def add_samples(self,\n",
    "                    file_paths):\n",
    "        \"\"\"\n",
    "        Method to initialize multiple sample objects based on file paths\n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        file_paths : dict \n",
    "            dictionary containing sample information \n",
    "        \"\"\"  \n",
    "\n",
    "        # Create a list to hold the sample objects\n",
    "        self.samples = []\n",
    "        for sample_name in file_paths:\n",
    "            attributes = file_paths[sample_name].copy()\n",
    "            attributes[\"sample_name\"] = sample_name # Assign sample name to the attributes\n",
    "            self.samples.append(self.init_sample(attributes)) # Add the sample to the list\n",
    "        return(file_paths)\n",
    "\n",
    "    def add_samples_from_samplesheet(self):\n",
    "        # Load file paths from samplesheet\n",
    "        file_paths = self.get_metadata_from_samplesheet()\n",
    "        # Add samples to samples list based on these paths\n",
    "        self.add_samples(file_paths)\n",
    "    \n",
    "    def add_samples_from_folder(self,\n",
    "                                input_config):\n",
    "        # Load file paths from specified folder\n",
    "        file_paths = self.get_input_from_folder(input_config)\n",
    "        # Add samples to samples list based on these paths\n",
    "        self.add_samples(file_paths)\n",
    "\n",
    "    def get_input_from_folder(self,\n",
    "                              input_config):\n",
    "        \"\"\"\n",
    "        Method to retrieve input files from a folder and classify them by type (e.g., Illumina or Nanopore) \n",
    "\n",
    "        Parameters\n",
    "        -------------\n",
    "        input_config : yaml file\n",
    "            Configuration file with options to handle samples\n",
    "        \"\"\" \n",
    "        # Regex for Illumina reads\n",
    "        illumina_regex = re.compile(input_config[\"file_patterns\"][\"Illumina_read_file\"])\n",
    "        # Regex for Nanopore reads\n",
    "        nanopore_regex = re.compile(input_config[\"file_patterns\"][\"Nanopore_read_file\"])\n",
    "        # Regex for assembly files\n",
    "        assembly_regex = re.compile(input_config[\"file_patterns\"][\"assembly_fasta_file\"])\n",
    "        \n",
    "        \n",
    "        input_folder = self.base_input_folder\n",
    "        file_paths = {} # Dict to store file paths by sample\n",
    "        files = os.listdir(input_folder)\n",
    "        \n",
    "        # Iterate over files\n",
    "        for file in files:\n",
    "            sample_name = False\n",
    "            if file.endswith((\".fa\",\".fasta\",\".fna\")): # Assembly data\n",
    "                # Extract sample name from filename\n",
    "                sample_name = re.match(assembly_regex,\n",
    "                                       file).group(\"sample_name\")\n",
    "                if sample_name in file_paths:\n",
    "                    file_paths[sample_name][\"assembly_file\"] = file\n",
    "                else:\n",
    "                    file_paths[sample_name] = {\"assembly_file\": file}\n",
    "            \n",
    "            # Reads\n",
    "            elif file.endswith(\".fastq.gz\"):\n",
    "                re_match = re.match(illumina_regex,file)\n",
    "                if re_match:\n",
    "                    # Get sample name\n",
    "                    sample_name = re_match.group(\"sample_name\")\n",
    "                    # Get read number\n",
    "                    read_number = re_match.group(\"paired_read_number\")\n",
    "                    \n",
    "                    # Distinguish between R1 and R2\n",
    "                    if read_number == \"R1\":  # Read 1\n",
    "                        if sample_name in file_paths:\n",
    "                            if \"Illumina_read_files\" in file_paths[sample_name]:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file + \",\" + file_paths[sample_name][\"Illumina_read_files\"]\n",
    "                            else:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Illumina_read_files\": file}\n",
    "                    elif read_number == \"R2\": # Read 2\n",
    "                        if sample_name in file_paths:\n",
    "                            if \"Illumina_read_files\" in file_paths[sample_name]:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file_paths[sample_name][\"Illumina_read_files\"] + \",\" + file\n",
    "                            else:\n",
    "                                file_paths[sample_name][\"Illumina_read_files\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Illumina_read_files\": file}\n",
    "                \n",
    "                #### TODO: Logic for nanopore read data, need file naming conventions\n",
    "                \n",
    "                # Nanopore reads if present\n",
    "                else:\n",
    "                    re_match = re.match(nanopore_regex,file)\n",
    "                    if re_match:\n",
    "                        sample_name = re_match.group(\"sample_name\")\n",
    "                        if sample_name in file_paths:\n",
    "                            file_paths[sample_name][\"Nanopore_read_file\"] = file\n",
    "                        else:\n",
    "                            file_paths[sample_name] = {\"Nanopore_read_file\": file}\n",
    "        \n",
    "        # Update attributes with metadata from samplesheet if provided\n",
    "        if self.metadata_file is not None:\n",
    "            for sample_name in file_paths:\n",
    "                if sample_name in self.metadata:\n",
    "                    file_paths_updated = self.metadata[sample_name].copy()\n",
    "                    file_paths_updated.update(file_paths[sample_name])\n",
    "                    file_paths[sample_name] = file_paths_updated\n",
    "        return(file_paths)\n",
    "    \n",
    "    def get_metadata_from_samplesheet(self):\n",
    "        # Load the samplesheet in a dataframe\n",
    "        df = core.get_samplesheet({\"path\": self.metadata_file})\n",
    "        df.set_index('sample_name', inplace=True)\n",
    "        df_dict = df.to_dict(\"index\")\n",
    "        return(df_dict)\n",
    "\n",
    "    def __iter__(self):\n",
    "        \"\"\"\n",
    "        Method to iterate over the samples in the input_manager object\n",
    "        \"\"\"\n",
    "        for sample in self.samples:\n",
    "            yield(sample)\n",
    "\n",
    "    def __len__(self):\n",
    "        \"\"\"\n",
    "        Method that returns the number of samples managed by this object\n",
    "        \"\"\"\n",
    "        return(len(self.samples))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "def get_abs_file_path(input_folder,\n",
    "                      file_path):\n",
    "    \"\"\"\n",
    "    Function to return the absolute path of a file\n",
    "    \"\"\"\n",
    "    if file_path is not None:\n",
    "        if input_folder:\n",
    "            file_path = os.path.abspath(os.path.join(input_folder,file_path))\n",
    "        else:\n",
    "            file_path = os.path.abspath(file_path)\n",
    "        if not os.path.exists(file_path):\n",
    "            print(f\"WARNING: {file_path} does not exist\")\n",
    "    return(file_path)\n",
    "\n",
    "\n",
    "def get_single_sample_attributes(input_config):\n",
    "    \"\"\"\n",
    "    Function to retrieve input information for a single sample.\n",
    "\n",
    "    Parameters\n",
    "    -------------\n",
    "    input_config : yaml file\n",
    "        Configuration file with options to handle samples\n",
    "    \"\"\" \n",
    "\n",
    "    # Get file paths from input config\n",
    "    file_paths = dict((k, input_config[k]) for k in [\"sample_name\",\n",
    "                                                     \"assembly_file\",\n",
    "                                                     \"Illumina_read_files\",\n",
    "                                                     \"Nanopore_read_file\"] if k in input_config)\n",
    "    # If the sample name is not found\n",
    "    if not \"sample_name\" in input_config or input_config[\"sample_name\"] is None:\n",
    "        # If the assembly file is available\n",
    "        if \"assembly_file\" in input_config and input_config[\"assembly_file\"] is not None:\n",
    "            # Add the sample name to the file path dictionary from the assembly file \n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"assembly_fasta_file\"],\n",
    "                                                 os.path.basename(input_config[\"assembly_file\"])).group(\"sample_name\")\n",
    "        # Otherwise take it from the Illumina reads\n",
    "        elif \"Illumina_read_files\" in input_config and input_config[\"Illumina_read_files\"] is not None:\n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"Illumina_read_file\"],\n",
    "                                                 os.path.basename(input_config[\"Illumina_read_files\"].split(',')[0])).group(\"sample_name\")\n",
    "        # Otherwise take it from the Nanopore reads\n",
    "        elif \"Nanopore_read_file\" in input_config and input_config[\"Nanopore_read_file\"] is not None:\n",
    "            file_paths[\"sample_name\"] = re.match(input_config[\"file_patterns\"][\"assembly_fasta_file\"],\n",
    "                                                 os.path.basename(input_config[\"Nanopore_read_file\"])).group(\"sample_name\")\n",
    "    \n",
    "    # If a samplesheet is available in the config\n",
    "    if input_config[\"samplesheet\"] is not None:\n",
    "        # Load sample sheet as DataFrame\n",
    "        df = core.get_samplesheet({\"path\": input_config[\"samplesheet\"]})\n",
    "        try:\n",
    "            # Set sample_name as the index if it exists\n",
    "            df.set_index('sample_name', inplace=True)\n",
    "            df_dict = df.to_dict(\"index\") # Save the sample bane\n",
    "            # Retrieve metadata for the specified sample name, if available\n",
    "            if input_config[\"sample_name\"] is not None and input_config[\"sample_name\"] in df_dict:\n",
    "                attributes = df_dict[input_config[\"sample_name\"]]\n",
    "            else:\n",
    "                # Otherwise, use the first row in the sample sheet as default metadata\n",
    "                attributes = df_dict[list(df_dict)[0]]\n",
    "                print(\"Sample name not provided or not found in metadata. Using first row as metadata input\")\n",
    "        except KeyError:\n",
    "            # If sample_name column is not found, fallback to using the first row of the DataFrame\n",
    "            df_list = df.to_dict(\"records\")\n",
    "            attributes = df_list[0]\n",
    "            print(\"No sample_name column in metadata. Using first row as metadata input\")\n",
    "    else:\n",
    "        attributes = {}\n",
    "    # Update attributes with file paths and sample name information\n",
    "    attributes.update(file_paths)\n",
    "    return(attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Python-dotenv could not parse statement starting at line 9\n",
      "Python-dotenv could not parse statement starting at line 10\n",
      "Python-dotenv could not parse statement starting at line 11\n",
      "Python-dotenv could not parse statement starting at line 12\n",
      "Python-dotenv could not parse statement starting at line 13\n",
      "Python-dotenv could not parse statement starting at line 16\n",
      "Python-dotenv could not parse statement starting at line 17\n",
      "Python-dotenv could not parse statement starting at line 18\n",
      "Python-dotenv could not parse statement starting at line 19\n",
      "Python-dotenv could not parse statement starting at line 22\n",
      "Python-dotenv could not parse statement starting at line 23\n",
      "Python-dotenv could not parse statement starting at line 24\n",
      "Python-dotenv could not parse statement starting at line 25\n",
      "Python-dotenv could not parse statement starting at line 26\n",
      "Python-dotenv could not parse statement starting at line 27\n",
      "Python-dotenv could not parse statement starting at line 28\n",
      "Python-dotenv could not parse statement starting at line 29\n",
      "Python-dotenv could not parse statement starting at line 30\n",
      "Python-dotenv could not parse statement starting at line 31\n",
      "Python-dotenv could not parse statement starting at line 35\n",
      "Python-dotenv could not parse statement starting at line 36\n",
      "Python-dotenv could not parse statement starting at line 37\n",
      "Python-dotenv could not parse statement starting at line 38\n",
      "Python-dotenv could not parse statement starting at line 39\n",
      "Python-dotenv could not parse statement starting at line 40\n",
      "Python-dotenv could not parse statement starting at line 42\n",
      "Python-dotenv could not parse statement starting at line 43\n",
      "Python-dotenv could not parse statement starting at line 45\n",
      "Python-dotenv could not parse statement starting at line 46\n",
      "Python-dotenv could not parse statement starting at line 48\n",
      "Python-dotenv could not parse statement starting at line 49\n",
      "Python-dotenv could not parse statement starting at line 50\n",
      "Python-dotenv could not parse statement starting at line 56\n",
      "Python-dotenv could not parse statement starting at line 57\n",
      "Python-dotenv could not parse statement starting at line 58\n",
      "Python-dotenv could not parse statement starting at line 59\n",
      "Python-dotenv could not parse statement starting at line 62\n",
      "Python-dotenv could not parse statement starting at line 63\n",
      "Python-dotenv could not parse statement starting at line 64\n",
      "Python-dotenv could not parse statement starting at line 65\n",
      "Python-dotenv could not parse statement starting at line 68\n",
      "Python-dotenv could not parse statement starting at line 69\n",
      "Python-dotenv could not parse statement starting at line 70\n",
      "Python-dotenv could not parse statement starting at line 71\n",
      "Python-dotenv could not parse statement starting at line 73\n",
      "Python-dotenv could not parse statement starting at line 76\n",
      "Python-dotenv could not parse statement starting at line 77\n",
      "Python-dotenv could not parse statement starting at line 78\n",
      "Python-dotenv could not parse statement starting at line 81\n",
      "Python-dotenv could not parse statement starting at line 82\n",
      "Python-dotenv could not parse statement starting at line 83\n",
      "Python-dotenv could not parse statement starting at line 86\n",
      "Python-dotenv could not parse statement starting at line 87\n",
      "Python-dotenv could not parse statement starting at line 88\n",
      "Python-dotenv could not parse statement starting at line 90\n",
      "Python-dotenv could not parse statement starting at line 91\n",
      "Python-dotenv could not parse statement starting at line 92\n",
      "Python-dotenv could not parse statement starting at line 98\n",
      "Python-dotenv could not parse statement starting at line 99\n",
      "Python-dotenv could not parse statement starting at line 100\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'file_patterns': {'Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"'}, 'load_from_samplesheet': True, 'load_from_folder': False, 'input_folder': None, 'output_folder': '.', 'assembly_file': None, 'Illumina_read_files': None, 'Nanopore_read_file': None, 'samplesheet': '/dpssi/home/scrsim/ssi_analysis_utility/examples/samplesheet.tsv', 'sample_name': None}\n",
      "{'base_input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'metadata_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/samplesheet.tsv', 'metadata': {'GAS-2022-1029': {'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'assembly_file': 'GAS-2022-1029.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'speC positive'}, 'GAS-2023-0253': {'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'Nan'}}, 'samples': [<__main__.sample_data object>, <__main__.sample_data object>]}\n"
     ]
    }
   ],
   "source": [
    "# |export\n",
    "config = core.get_config(\"/dpssi/home/scrsim/ssi_analysis_utility/ssi_analysis_utility/config/config.default.yaml\")\n",
    "simone_testing = input_manager(config[\"input_manager\"])\n",
    "print(simone_testing.__dict__)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# |export\n",
    "\n",
    "def single_test():\n",
    "    example_sample = sample_data({\"sample_name\":\"GAS-2022-1029\",\n",
    "                                       \"assembly_file\":\"examples/GAS-2022-1029.fasta\",\n",
    "                                       \"Illumina_read_files\":[\"examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz\",\"examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz\"]},\n",
    "                                       input_folder = False,\n",
    "                                       )\n",
    "    assert(len(example_sample.Illumina_read_files) == 2)\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "def single_test_assembly():\n",
    "    example_sample = sample_data({\"assembly_file\":\"examples/GAS-2022-1029.fasta\"},\n",
    "                                       input_folder = False,\n",
    "                                       )\n",
    "    assert(not(example_sample.Illumina_read_files))\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "def single_test_input_folder():\n",
    "    config = core.get_config()\n",
    "    example_sample = sample_data({\"sample_name\":\"GAS-2024-0773\",\n",
    "                                       \"assembly_file\":\"GAS-2024-0773.fasta\",\n",
    "                                       \"Illumina_read_files\":[\"GAS-2024-0773_S35_L555_R1_001.fastq.gz\",\"GAS-2024-0773_S35_L555_R2_001.fastq.gz\"]},\n",
    "                                       input_folder = \"examples\",\n",
    "                                       )\n",
    "    assert(example_sample.sample_name == \"GAS-2024-0773\")\n",
    "    assert(len(example_sample.Illumina_read_files) == 2)\n",
    "    assert(not example_sample.Nanopore_read_file)\n",
    "    print(example_sample.__dict__)\n",
    "\n",
    "\n",
    "def test_folder():\n",
    "    config = core.get_config()\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_folder\"] = True\n",
    "    input_config[\"input_folder\"] = \"examples/\"\n",
    "    test = input_manager(input_config)\n",
    "    for x in test:\n",
    "        print(x.__dict__)\n",
    "\n",
    "def test_folder_with_metadata():\n",
    "    config = core.get_config()\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_folder\"] = True\n",
    "    input_config[\"input_folder\"] = \"examples/\"\n",
    "    input_config[\"samplesheet\"] = \"examples/samplesheet.tsv\"\n",
    "    test = input_manager(input_config)\n",
    "    for x in test:\n",
    "        print(x.__dict__)\n",
    "\n",
    "def samplesheet_test():\n",
    "    config = core.get_config()\n",
    "    print(config)\n",
    "    input_config =  config[\"input_manager\"]\n",
    "    input_config[\"load_from_samplesheet\"] = True\n",
    "    input_config[\"samplesheet\"] = \"/dpssi/home/scrsim/ssi_analysis_utility/examples/samplesheet.tsv\"\n",
    "    test = input_manager(input_config)\n",
    "    print(test.__dict__)\n",
    "\n",
    "    # for x in test.__dict__:\n",
    "    #     print(x)\n",
    "    #     print(x.metadata[\"Illumina_read_files\"])\n",
    "    #     print(x.metadata[\"assembly_file\"])\n",
    "    #     print(x.metadata[\"sample_name\"])\n",
    "    #     print(x.assembly_file)\n",
    "    #     print(x.Illumina_read_files)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': None, 'sample_name': 'GAS-2022-1029', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'sample_name': 'GAS-2022-1029', 'assembly_file': 'examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', 'examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz']}}\n",
      "{'input_folder': None, 'sample_name': 'sample', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'examples/GAS-2022-1029.fasta'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'sample_name': 'GAS-2024-0773', 'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': ['GAS-2024-0773_S35_L555_R1_001.fastq.gz', 'GAS-2024-0773_S35_L555_R2_001.fastq.gz']}}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "single_test()\n",
    "single_test_assembly()\n",
    "single_test_input_folder()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2023-0253', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_S74_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_S74_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_nanopore.fastq.gz', 'metadata': {'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'sample_name': 'GAS-2023-0253'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2022-1029', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_nanopore.fastq.gz', 'metadata': {'assembly_file': 'GAS-2022-1029.fasta', 'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'sample_name': 'GAS-2022-1029'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': 'GAS-2024-0773_S35_L555_R1_001.fastq.gz,GAS-2024-0773_S35_L555_R2_001.fastq.gz', 'sample_name': 'GAS-2024-0773'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2024-0665', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0665.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0665.fasta', 'sample_name': 'GAS-2024-0665'}}\n"
     ]
    }
   ],
   "source": [
    "test_folder()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2023-0253', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_S74_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_S74_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2023-0253_nanopore.fastq.gz', 'metadata': {'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'Nan', 'sample_name': 'GAS-2023-0253'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2022-1029', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_S42_L555_R2_001.fastq.gz'], 'Nanopore_read_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2022-1029_nanopore.fastq.gz', 'metadata': {'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'assembly_file': 'GAS-2022-1029.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'speC positive', 'sample_name': 'GAS-2022-1029'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2024-0773', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773.fasta', 'Illumina_read_files': ['/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R1_001.fastq.gz', '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0773_S35_L555_R2_001.fastq.gz'], 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0773.fasta', 'Illumina_read_files': 'GAS-2024-0773_S35_L555_R1_001.fastq.gz,GAS-2024-0773_S35_L555_R2_001.fastq.gz', 'sample_name': 'GAS-2024-0773'}}\n",
      "{'input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'sample_name': 'GAS-2024-0665', 'assembly_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/GAS-2024-0665.fasta', 'Illumina_read_files': None, 'Nanopore_read_file': None, 'metadata': {'assembly_file': 'GAS-2024-0665.fasta', 'sample_name': 'GAS-2024-0665'}}\n"
     ]
    }
   ],
   "source": [
    "test_folder_with_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'SHELL': '/bin/bash', 'HISTCONTROL': 'erasedups', 'CONDA_EXE': '/users/data/Tools/Conda/Miniconda3-py311_23.5.2-0-Linux-x86_64/bin/conda', '_CE_M': '', 'HISTSIZE': '10000', 'HISTTIMEFORMAT': '[%F %T] ', 'XML_CATALOG_FILES': 'file:///dpssi/home/scrsim/conda_envs/.venv/etc/xml/catalog file:///etc/xml/catalog', 'ENV': '/cm/local/apps/environment-modules/4.2.1//init/profile.sh', 'PWD': '/dpssi/home/scrsim', 'KRB5CCNAME': 'KCM:', 'LOGNAME': 'scrsim', 'XDG_SESSION_TYPE': 'tty', 'CONDA_PREFIX': '/dpssi/home/scrsim/conda_envs/.venv', 'MODULESHOME': '/cm/local/apps/environment-modules/4.2.1/', 'MOTD_SHOWN': 'pam', 'HOME': '/dpssi/home/scrsim', 'LANG': 'en_US.UTF-8', 'CONDA_PROMPT_MODIFIER': '(/dpssi/home/scrsim/conda_envs/.venv) ', 'VSCODE_AGENT_FOLDER': '/dpssi/home/scrsim/.vscode-server', 'SSH_CONNECTION': '10.96.143.48 50049 10.45.128.75 22', 'XDG_SESSION_CLASS': 'user', '_CE_CONDA': '', 'LESSOPEN': '||/usr/bin/lesspipe.sh %s', 'USER': 'scrsim', 'CONDA_SHLVL': '2', 'VISUAL': 'nano', 'LOADEDMODULES': '', 'SHLVL': '2', 'BASH_ENV': '/cm/local/apps/environment-modules/4.2.1//init/bash', 'XDG_SESSION_ID': '1995', 'CONDA_PYTHON_EXE': '/users/data/Tools/Conda/Miniconda3-py311_23.5.2-0-Linux-x86_64/bin/python', 'XDG_RUNTIME_DIR': '/run/user/7200225', 'S_COLORS': 'auto', 'SSH_CLIENT': '10.96.143.48 50049 22', 'CONDA_DEFAULT_ENV': '/dpssi/home/scrsim/conda_envs/.venv', 'DEBUGINFOD_URLS': 'https://debuginfod.centos.org/ ', 'ENABLE_LMOD': '0', 'which_declare': 'declare -f', 'KDEDIRS': '/usr', 'XDG_DATA_DIRS': '/dpssi/home/scrsim/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share', 'PATH': '/dpssi/home/scrsim/conda_envs/.venv/bin:/cm/local/apps/environment-modules/4.2.1/bin:/dpssi/home/scrsim/.vscode-server/bin/f1b07bd25dfad64b0167beb15359ae573aecd2cc/bin/remote-cli:/users/data/Tools/bin:/cm/local/apps/environment-modules/4.2.1/bin:/dpssi/home/scrsim/conda_envs/.venv/bin:/users/data/Tools/Conda/Miniconda3-py311_23.5.2-0-Linux-x86_64/condabin:/users/data/Tools/bin:/cm/local/apps/environment-modules/4.2.1/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/sbin:/cm/local/apps/environment-modules/4.2.1/bin:/dpssi/home/scrsim/.local/bin:/dpssi/home/scrsim/bin', 'MODULEPATH': '/etc/scl/modulefiles:/etc/scl/modulefiles:/cm/local/modulefiles:/cm/local/.modulefiles_cache/shared/modulefiles', 'HISTFILESIZE': '10000', 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/7200225/bus', 'MODULES_CMD': '/cm/local/apps/environment-modules/4.2.1/libexec/modulecmd.tcl', 'BASH_FUNC_switchml%%': '() {  typeset swfound=1;\\n if [ \"${MODULES_USE_COMPAT_VERSION:-0}\" = \\'1\\' ]; then\\n typeset swname=\\'main\\';\\n if [ -e /cm/local/apps/environment-modules/4.2.1//libexec/modulecmd.tcl ]; then\\n typeset swfound=0;\\n unset MODULES_USE_COMPAT_VERSION;\\n fi;\\n else\\n typeset swname=\\'compatibility\\';\\n if [ -e /cm/local/apps/environment-modules/4.2.1//libexec/modulecmd-compat ]; then\\n typeset swfound=0;\\n MODULES_USE_COMPAT_VERSION=1;\\n export MODULES_USE_COMPAT_VERSION;\\n fi;\\n fi;\\n if [ $swfound -eq 0 ]; then\\n echo \"Switching to Modules $swname version\";\\n source /cm/local/apps/environment-modules/4.2.1//init/bash;\\n else\\n echo \"Cannot switch to Modules $swname version, command not found\";\\n return 1;\\n fi\\n}', 'BASH_FUNC_which%%': '() {  ( alias;\\n eval ${which_declare} ) | /usr/bin/which --tty-only --read-alias --read-functions --show-tilde --show-dot $@\\n}', 'BASH_FUNC_module%%': '() {  unset _mlshdbg;\\n if [ \"${MODULES_SILENT_SHELL_DEBUG:-0}\" = \\'1\\' ]; then\\n case \"$-\" in \\n *v*x*)\\n set +vx;\\n _mlshdbg=\\'vx\\'\\n ;;\\n *v*)\\n set +v;\\n _mlshdbg=\\'v\\'\\n ;;\\n *x*)\\n set +x;\\n _mlshdbg=\\'x\\'\\n ;;\\n *)\\n _mlshdbg=\\'\\'\\n ;;\\n esac;\\n fi;\\n unset _mlre _mlIFS;\\n if [ -n \"${IFS+x}\" ]; then\\n _mlIFS=$IFS;\\n fi;\\n IFS=\\' \\';\\n for _mlv in ${MODULES_RUN_QUARANTINE:-};\\n do\\n if [ \"${_mlv}\" = \"${_mlv##*[!A-Za-z0-9_]}\" -a \"${_mlv}\" = \"${_mlv#[0-9]}\" ]; then\\n if [ -n \"`eval \\'echo ${\\'$_mlv\\'+x}\\'`\" ]; then\\n _mlre=\"${_mlre:-}${_mlv}_modquar=\\'`eval \\'echo ${\\'$_mlv\\'}\\'`\\' \";\\n fi;\\n _mlrv=\"MODULES_RUNENV_${_mlv}\";\\n _mlre=\"${_mlre:-}${_mlv}=\\'`eval \\'echo ${\\'$_mlrv\\':-}\\'`\\' \";\\n fi;\\n done;\\n if [ -n \"${_mlre:-}\" ]; then\\n eval `eval ${_mlre}/usr/bin/tclsh /cm/local/apps/environment-modules/4.2.1/libexec/modulecmd.tcl bash \\'\"$@\"\\'`;\\n else\\n eval `/usr/bin/tclsh /cm/local/apps/environment-modules/4.2.1/libexec/modulecmd.tcl bash \"$@\"`;\\n fi;\\n _mlstatus=$?;\\n if [ -n \"${_mlIFS+x}\" ]; then\\n IFS=$_mlIFS;\\n else\\n unset IFS;\\n fi;\\n unset _mlre _mlv _mlrv _mlIFS;\\n if [ -n \"${_mlshdbg:-}\" ]; then\\n set -$_mlshdbg;\\n fi;\\n unset _mlshdbg;\\n return $_mlstatus\\n}', 'BASH_FUNC_scl%%': '() {  if [ \"$1\" = \"load\" -o \"$1\" = \"unload\" ]; then\\n eval \"module $@\";\\n else\\n /usr/bin/scl \"$@\";\\n fi\\n}', '_': '/dpssi/home/scrsim/conda_envs/.venv/bin/python', 'VSCODE_HANDLES_SIGPIPE': 'true', 'HOSTNAME': 'dpssi-002.computerome.local', 'PS1': '(base) \\\\[\\\\e[1;36m\\\\]\\\\u\\\\[\\\\e[31m\\\\]@\\\\[\\\\e[1;36m\\\\]\\\\h\\\\[\\\\e[31m\\\\]:\\\\[\\\\e[1;34m\\\\]\\\\w \\\\[\\\\e[31m\\\\]>\\\\[\\\\e[0m\\\\] ', 'MAIL': '/var/spool/mail/scrsim', 'VSCODE_AMD_ENTRYPOINT': 'vs/workbench/api/node/extensionHostProcess', 'VSCODE_HANDLES_UNCAUGHT_ERRORS': 'true', 'VSCODE_NLS_CONFIG': '{\"locale\":\"en\",\"osLocale\":\"en\",\"availableLanguages\":{}}', 'BROWSER': '/dpssi/home/scrsim/.vscode-server/bin/f1b07bd25dfad64b0167beb15359ae573aecd2cc/bin/helpers/browser.sh', 'VSCODE_CWD': '/dpssi/home/scrsim', 'ELECTRON_RUN_AS_NODE': '1', 'VSCODE_IPC_HOOK_CLI': '/run/user/7200225/vscode-ipc-8df73ba4-63fb-4a83-8af7-660ef537498a.sock', 'PYTHONUNBUFFERED': '1', 'GSETTINGS_SCHEMA_DIR': '/dpssi/home/scrsim/conda_envs/.venv/share/glib-2.0/schemas', 'CONDA_ROOT': '/users/data/Tools/Conda/Miniconda3-py311_23.5.2-0-Linux-x86_64', 'DENO_DOM_VERSION': '0.1.35-alpha', 'GSETTINGS_SCHEMA_DIR_CONDA_BACKUP': '', 'QUARTO_CONDA_PREFIX': '/dpssi/home/scrsim/conda_envs/.venv', 'DENO_DOM_PLUGIN': '/dpssi/home/scrsim/conda_envs/.venv/lib/deno_dom.so', 'DENO_INSTALL_ROOT': '/dpssi/home/scrsim/conda_envs/.venv', 'PYTHONIOENCODING': 'utf-8', 'QUARTO_SHARE_PATH': '/dpssi/home/scrsim/conda_envs/.venv/share/quarto', 'QUARTO_DENO': '/dpssi/home/scrsim/conda_envs/.venv/bin/deno', 'QUARTO_DART_SASS': '/dpssi/home/scrsim/conda_envs/.venv/bin/sass', 'QUARTO_PANDOC': '/dpssi/home/scrsim/conda_envs/.venv/bin/pandoc', 'QUARTO_TYPST': '/dpssi/home/scrsim/conda_envs/.venv/bin/typst', 'QUARTO_DENO_DOM': '/dpssi/home/scrsim/conda_envs/.venv/lib/deno_dom.so', 'QUARTO_ESBUILD': '/dpssi/home/scrsim/conda_envs/.venv/bin/esbuild', 'CONDA_PREFIX_1': '/users/data/Tools/Conda/Miniconda3-py311_23.5.2-0-Linux-x86_64', 'PYDEVD_IPYTHON_COMPATIBLE_DEBUGGING': '1', 'PYDEVD_USE_FRAME_EVAL': 'NO', 'TERM': 'xterm-color', 'CLICOLOR': '1', 'FORCE_COLOR': '1', 'CLICOLOR_FORCE': '1', 'PAGER': 'cat', 'GIT_PAGER': 'cat', 'MPLBACKEND': 'module://matplotlib_inline.backend_inline', 'CORE_PROJECT_VARIABLE_PREFIX': 'STREPTOFILE_', 'CORE_TEST_VAR': 'Test', 'STREPTOFILE_RESOURCES_DIR': './resources', 'input_parameters': {'sample_name': 'sample_name', 'assembly_file': 'assembly_file', 'Illumina_read_files': 'Illumina_read_files', 'Nanopore_read_file': 'Nanopore_read_file'}, 'input_parameters.sample_name': 'sample_name', 'input_parameters.assembly_file': 'assembly_file', 'input_parameters.Illumina_read_files': 'Illumina_read_files', 'input_parameters.Nanopore_read_file': 'Nanopore_read_file', 'file_patterns': {'Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"'}, 'file_patterns.Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'file_patterns.Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'file_patterns.assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'file_patterns.metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"', 'input_manager': {'file_patterns': {'Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"'}, 'load_from_samplesheet': False, 'load_from_folder': False, 'input_folder': None, 'output_folder': '.', 'assembly_file': None, 'Illumina_read_files': None, 'Nanopore_read_file': None, 'samplesheet': None, 'sample_name': None}, 'input_manager.file_patterns': {'Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"'}, 'input_manager.file_patterns.Illumina_read_file': '(?P<sample_name>.+?)(?P<sample_number>(_S[0-9]+)?)(?P<lane>(_L[0-9]+)?)_(?P<paired_read_number>R[1|2])(?P<set_number>(_[0-9]+)?)(?P<file_extension>\\\\.fastq\\\\.gz)', 'input_manager.file_patterns.Nanopore_read_file': '(?P<sample_name>.+?)_nanopore\\\\.fastq\\\\.gz', 'input_manager.file_patterns.assembly_fasta_file': '(?P<sample_name>.+?)(\\\\.fa|\\\\.fna|\\\\.fasta)', 'input_manager.file_patterns.metadata': 'r\"(?P<metadata_name>.+?)(\\\\.tsv|\\\\.tab\\\\.txt)\"', 'input_manager.load_from_samplesheet': False, 'input_manager.load_from_folder': False, 'input_manager.input_folder': None, 'input_manager.output_folder': '.', 'input_manager.assembly_file': None, 'input_manager.Illumina_read_files': None, 'input_manager.Nanopore_read_file': None, 'input_manager.samplesheet': None, 'input_manager.sample_name': None, 'analysis_settings': {'blast_presence_absence': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast']}, 'assembly_lineage_determination': {'alias': 'lineage_determination', 'percent_snp_threshold': 50, 'files_to_clean': ['delta', 'filtered_delta', 'frankenfasta']}, 'Spyogenes': {'analyses_to_run': {'emm_typing': 'Typing of M protein gene using blast against CDC curated alleles', 'assembly_lineage_determination': 'Lineage determination based on presence of specified SNPs when mapping genome assembly against reference genome', 'resistance_gene_detection': 'Presence of resistance gens tetM, ermA and ermB', 'virulence_gene_detection': 'Presence of Streptococcal virulence genes from the Virulence Factor Database'}, 'assembly_lineage_determination': {'alias': 'assembly_lineage_determination', 'percent_snp_threshold': 50, 'files_to_clean': ['delta', 'filtered_delta', 'frankenfasta'], 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv'}, 'emm_typing': {'alias': 'emm_typing', 'emm_allele_file': './resources/emm_typing/emm_alleles.fasta', 'emm_cluster_file': './resources/emm_typing/emm_enn_mrp_subgroups.txt', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'files_to_clean': ['blast']}, 'resistance_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'resistance_genes', 'query_fasta_file': './resources/AMR/AMR_genes.fasta'}, 'virulence_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'virulence_genes', 'query_fasta_file': './resources/virulence_factors/VFDB_genes.fasta'}, 'Illumina_lineage_determination': {'alias': 'Illumina_lineage_determination', 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv', 'files_to_clean': ['vcf', 'bam', 'bam_index']}}}, 'analysis_settings.blast_presence_absence': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast']}, 'analysis_settings.blast_presence_absence.pident_threshold': 80, 'analysis_settings.blast_presence_absence.cov_threshold': 80, 'analysis_settings.blast_presence_absence.additional_blast_parameters': '', 'analysis_settings.blast_presence_absence.blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'analysis_settings.blast_presence_absence.results_format': 'matrix', 'analysis_settings.blast_presence_absence.files_to_clean': ['blast'], 'analysis_settings.blast_presence_absence.files_to_clean.0': 'blast', 'analysis_settings.assembly_lineage_determination': {'alias': 'lineage_determination', 'percent_snp_threshold': 50, 'files_to_clean': ['delta', 'filtered_delta', 'frankenfasta']}, 'analysis_settings.assembly_lineage_determination.alias': 'lineage_determination', 'analysis_settings.assembly_lineage_determination.percent_snp_threshold': 50, 'analysis_settings.assembly_lineage_determination.files_to_clean': ['delta', 'filtered_delta', 'frankenfasta'], 'analysis_settings.assembly_lineage_determination.files_to_clean.0': 'delta', 'analysis_settings.assembly_lineage_determination.files_to_clean.1': 'filtered_delta', 'analysis_settings.assembly_lineage_determination.files_to_clean.2': 'frankenfasta', 'analysis_settings.Spyogenes': {'analyses_to_run': {'emm_typing': 'Typing of M protein gene using blast against CDC curated alleles', 'assembly_lineage_determination': 'Lineage determination based on presence of specified SNPs when mapping genome assembly against reference genome', 'resistance_gene_detection': 'Presence of resistance gens tetM, ermA and ermB', 'virulence_gene_detection': 'Presence of Streptococcal virulence genes from the Virulence Factor Database'}, 'assembly_lineage_determination': {'alias': 'assembly_lineage_determination', 'percent_snp_threshold': 50, 'files_to_clean': ['delta', 'filtered_delta', 'frankenfasta'], 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv'}, 'emm_typing': {'alias': 'emm_typing', 'emm_allele_file': './resources/emm_typing/emm_alleles.fasta', 'emm_cluster_file': './resources/emm_typing/emm_enn_mrp_subgroups.txt', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'files_to_clean': ['blast']}, 'resistance_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'resistance_genes', 'query_fasta_file': './resources/AMR/AMR_genes.fasta'}, 'virulence_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'virulence_genes', 'query_fasta_file': './resources/virulence_factors/VFDB_genes.fasta'}, 'Illumina_lineage_determination': {'alias': 'Illumina_lineage_determination', 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv', 'files_to_clean': ['vcf', 'bam', 'bam_index']}}, 'analysis_settings.Spyogenes.analyses_to_run': {'emm_typing': 'Typing of M protein gene using blast against CDC curated alleles', 'assembly_lineage_determination': 'Lineage determination based on presence of specified SNPs when mapping genome assembly against reference genome', 'resistance_gene_detection': 'Presence of resistance gens tetM, ermA and ermB', 'virulence_gene_detection': 'Presence of Streptococcal virulence genes from the Virulence Factor Database'}, 'analysis_settings.Spyogenes.analyses_to_run.emm_typing': 'Typing of M protein gene using blast against CDC curated alleles', 'analysis_settings.Spyogenes.analyses_to_run.assembly_lineage_determination': 'Lineage determination based on presence of specified SNPs when mapping genome assembly against reference genome', 'analysis_settings.Spyogenes.analyses_to_run.resistance_gene_detection': 'Presence of resistance gens tetM, ermA and ermB', 'analysis_settings.Spyogenes.analyses_to_run.virulence_gene_detection': 'Presence of Streptococcal virulence genes from the Virulence Factor Database', 'analysis_settings.Spyogenes.assembly_lineage_determination': {'alias': 'assembly_lineage_determination', 'percent_snp_threshold': 50, 'files_to_clean': ['delta', 'filtered_delta', 'frankenfasta'], 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv'}, 'analysis_settings.Spyogenes.assembly_lineage_determination.alias': 'assembly_lineage_determination', 'analysis_settings.Spyogenes.assembly_lineage_determination.percent_snp_threshold': 50, 'analysis_settings.Spyogenes.assembly_lineage_determination.files_to_clean': ['delta', 'filtered_delta', 'frankenfasta'], 'analysis_settings.Spyogenes.assembly_lineage_determination.files_to_clean.0': 'delta', 'analysis_settings.Spyogenes.assembly_lineage_determination.files_to_clean.1': 'filtered_delta', 'analysis_settings.Spyogenes.assembly_lineage_determination.files_to_clean.2': 'frankenfasta', 'analysis_settings.Spyogenes.assembly_lineage_determination.reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'analysis_settings.Spyogenes.assembly_lineage_determination.lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv', 'analysis_settings.Spyogenes.emm_typing': {'alias': 'emm_typing', 'emm_allele_file': './resources/emm_typing/emm_alleles.fasta', 'emm_cluster_file': './resources/emm_typing/emm_enn_mrp_subgroups.txt', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'files_to_clean': ['blast']}, 'analysis_settings.Spyogenes.emm_typing.alias': 'emm_typing', 'analysis_settings.Spyogenes.emm_typing.emm_allele_file': './resources/emm_typing/emm_alleles.fasta', 'analysis_settings.Spyogenes.emm_typing.emm_cluster_file': './resources/emm_typing/emm_enn_mrp_subgroups.txt', 'analysis_settings.Spyogenes.emm_typing.blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'analysis_settings.Spyogenes.emm_typing.files_to_clean': ['blast'], 'analysis_settings.Spyogenes.emm_typing.files_to_clean.0': 'blast', 'analysis_settings.Spyogenes.resistance_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'resistance_genes', 'query_fasta_file': './resources/AMR/AMR_genes.fasta'}, 'analysis_settings.Spyogenes.resistance_gene_detection.pident_threshold': 80, 'analysis_settings.Spyogenes.resistance_gene_detection.cov_threshold': 80, 'analysis_settings.Spyogenes.resistance_gene_detection.additional_blast_parameters': '', 'analysis_settings.Spyogenes.resistance_gene_detection.blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'analysis_settings.Spyogenes.resistance_gene_detection.results_format': 'matrix', 'analysis_settings.Spyogenes.resistance_gene_detection.files_to_clean': ['blast'], 'analysis_settings.Spyogenes.resistance_gene_detection.files_to_clean.0': 'blast', 'analysis_settings.Spyogenes.resistance_gene_detection.alias': 'resistance_genes', 'analysis_settings.Spyogenes.resistance_gene_detection.query_fasta_file': './resources/AMR/AMR_genes.fasta', 'analysis_settings.Spyogenes.virulence_gene_detection': {'pident_threshold': 80, 'cov_threshold': 80, 'additional_blast_parameters': '', 'blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'results_format': 'matrix', 'files_to_clean': ['blast'], 'alias': 'virulence_genes', 'query_fasta_file': './resources/virulence_factors/VFDB_genes.fasta'}, 'analysis_settings.Spyogenes.virulence_gene_detection.pident_threshold': 80, 'analysis_settings.Spyogenes.virulence_gene_detection.cov_threshold': 80, 'analysis_settings.Spyogenes.virulence_gene_detection.additional_blast_parameters': '', 'analysis_settings.Spyogenes.virulence_gene_detection.blast_header': 'qseqid sseqid pident length qlen qstart qend sstart send sseq evalue bitscore', 'analysis_settings.Spyogenes.virulence_gene_detection.results_format': 'matrix', 'analysis_settings.Spyogenes.virulence_gene_detection.files_to_clean': ['blast'], 'analysis_settings.Spyogenes.virulence_gene_detection.files_to_clean.0': 'blast', 'analysis_settings.Spyogenes.virulence_gene_detection.alias': 'virulence_genes', 'analysis_settings.Spyogenes.virulence_gene_detection.query_fasta_file': './resources/virulence_factors/VFDB_genes.fasta', 'analysis_settings.Spyogenes.Illumina_lineage_determination': {'alias': 'Illumina_lineage_determination', 'reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv', 'files_to_clean': ['vcf', 'bam', 'bam_index']}, 'analysis_settings.Spyogenes.Illumina_lineage_determination.alias': 'Illumina_lineage_determination', 'analysis_settings.Spyogenes.Illumina_lineage_determination.reference_fasta_file': './resources/lineage_determination/MGAS5005.fasta', 'analysis_settings.Spyogenes.Illumina_lineage_determination.lineage_variant_file': './resources/lineage_determination/Spyogenes_LOCs.tsv', 'analysis_settings.Spyogenes.Illumina_lineage_determination.files_to_clean': ['vcf', 'bam', 'bam_index'], 'analysis_settings.Spyogenes.Illumina_lineage_determination.files_to_clean.0': 'vcf', 'analysis_settings.Spyogenes.Illumina_lineage_determination.files_to_clean.1': 'bam', 'analysis_settings.Spyogenes.Illumina_lineage_determination.files_to_clean.2': 'bam_index', 'databases': {'emm_typing': {'local_path': './resources/emm_typing/emm_alleles.fasta', 'online_path': 'https://ftp.cdc.gov/pub/infectious_diseases/biotech/tsemm/alltrimmed.tfa', 'update_protocol': 'wget'}}, 'databases.emm_typing': {'local_path': './resources/emm_typing/emm_alleles.fasta', 'online_path': 'https://ftp.cdc.gov/pub/infectious_diseases/biotech/tsemm/alltrimmed.tfa', 'update_protocol': 'wget'}, 'databases.emm_typing.local_path': './resources/emm_typing/emm_alleles.fasta', 'databases.emm_typing.online_path': 'https://ftp.cdc.gov/pub/infectious_diseases/biotech/tsemm/alltrimmed.tfa', 'databases.emm_typing.update_protocol': 'wget'}\n",
      "{'base_input_folder': '/dpssi/home/scrsim/ssi_analysis_utility/examples', 'metadata_file': '/dpssi/home/scrsim/ssi_analysis_utility/examples/samplesheet.tsv', 'metadata': {'GAS-2022-1029': {'Illumina_read_files': 'GAS-2022-1029_S42_L555_R1_001.fastq.gz,GAS-2022-1029_S42_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2022-1029_nanopore.fastq.gz', 'assembly_file': 'GAS-2022-1029.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'speC positive'}, 'GAS-2023-0253': {'Illumina_read_files': 'GAS-2023-0253_S74_L555_R1_001.fastq.gz,GAS-2023-0253_S74_L555_R2_001.fastq.gz', 'Nanopore_read_file': 'GAS-2023-0253_nanopore.fastq.gz', 'assembly_file': 'GAS-2023-0253.fasta', 'organism': 'Streptococcus pyogenes', 'variant': 'M1DK', 'notes': 'Nan'}}, 'samples': [<__main__.sample_data object>, <__main__.sample_data object>]}\n"
     ]
    }
   ],
   "source": [
    "samplesheet_test()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "# This is included at the end to ensure when you run through your notebook the code is also transferred to the module and isn't just a notebook\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
